File: src/hamiltonian.js
/**
 * Convierte el bloque editable en una lista de vecinos. El grafo se trata como
 * no dirigido porque el Avance 1 propone usar el mismo peso en ambos sentidos.
 */
export function buildAdjacency(block) {
  const adjacency = new Map(block.nodes.map((node) => [node.id, []]));
  for (const edge of block.edges) {
    if (!adjacency.has(edge.from) || !adjacency.has(edge.to) || edge.from === edge.to) continue;
    const weight = Number(edge.weight);
    if (!Number.isFinite(weight) || weight <= 0) continue;
    adjacency.get(edge.from).push({ id: edge.to, weight });
    adjacency.get(edge.to).push({ id: edge.from, weight });
  }
  return adjacency;
}

export function isConnected(adjacency, start) {
  if (!adjacency.size) return false;
  const visited = new Set([start]);
  const pending = [start];
  while (pending.length) {
    const current = pending.pop();
    for (const neighbor of adjacency.get(current) || []) {
      if (!visited.has(neighbor.id)) {
        visited.add(neighbor.id);
        pending.push(neighbor.id);
      }
    }
  }
  return visited.size === adjacency.size;
}

/**
 * Evalúa Dirac y Ore como criterios suficientes. Aunque uno no se cumpla,
 * el algoritmo de búsqueda aún puede encontrar un circuito hamiltoniano.
 */
export function evaluateHamiltonianCriteria(adjacency) {
  const ids = [...adjacency.keys()];
  const n = ids.length;
  const degrees = Object.fromEntries(ids.map((id) => [id, (adjacency.get(id) || []).length]));
  const connected = n > 0 && isConnected(adjacency, ids[0]);
  const minDegree = n ? Math.min(...Object.values(degrees)) : 0;
  const dirac = n >= 3 && minDegree >= n / 2;
  let ore = n >= 3;
  for (let index = 0; index < ids.length; index += 1) {
    for (let other = index + 1; other < ids.length; other += 1) {
      const adjacent = (adjacency.get(ids[index]) || []).some((neighbor) => neighbor.id === ids[other]);
      if (!adjacent && degrees[ids[index]] + degrees[ids[other]] < n) ore = false;
    }
  }
  return { n, connected, degrees, minDegree, dirac, ore, minimumDegree: minDegree >= 2 };
}

/**
 * Búsqueda con retroceso. No usa un recorrido fijo: explora las conexiones
 * que el usuario haya creado y conserva el ciclo válido de menor peso total.
 */
export function findHamiltonianCycle(block, startId) {
  const adjacency = buildAdjacency(block);
  const ids = [...adjacency.keys()];
  if (!ids.includes(startId) || ids.length < 3) return null;

  let best = null;
  const visited = new Set([startId]);
  const path = [startId];

  function search(current, totalWeight) {
    if (path.length === ids.length) {
      const closingEdge = (adjacency.get(current) || []).find((neighbor) => neighbor.id === startId);
      if (closingEdge) {
        const candidate = { path: [...path, startId], totalWeight: totalWeight + closingEdge.weight };
        if (!best || candidate.totalWeight < best.totalWeight) best = candidate;
      }
      return;
    }

    // Se prueban primero aristas cortas para encontrar pronto una buena solución.
    const candidates = [...(adjacency.get(current) || [])]
      .filter((neighbor) => !visited.has(neighbor.id))
      .sort((a, b) => a.weight - b.weight);
    for (const neighbor of candidates) {
      if (best && totalWeight + neighbor.weight >= best.totalWeight) continue;
      visited.add(neighbor.id);
      path.push(neighbor.id);
      search(neighbor.id, totalWeight + neighbor.weight);
      path.pop();
      visited.delete(neighbor.id);
    }
  }

  search(startId, 0);
  return best;
}

File: hamiltonian.py
"""Algoritmo de circuitos hamiltonianos y criterios de la Entrega 2."""

from __future__ import annotations

from math import inf


def build_adjacency(block: dict) -> dict[str, list[tuple[str, float]]]:
    """Construye un grafo no dirigido desde las aristas editables del bloque."""
    adjacency = {node["id"]: [] for node in block.get("nodes", [])}
    for edge in block.get("edges", []):
        origin, destination = edge.get("from"), edge.get("to")
        weight = float(edge.get("weight", 0))
        # Dos sucursales pueden estar dentro del mismo centro comercial y compartir
        # coordenadas. Una arista de 0 km sigue siendo válida entre nodos distintos.
        if origin not in adjacency or destination not in adjacency or origin == destination or weight < 0:
            continue
        adjacency[origin].append((destination, weight))
        adjacency[destination].append((origin, weight))
    return adjacency


def is_connected(adjacency: dict[str, list[tuple[str, float]]]) -> bool:
    """Verifica conectividad, condición básica antes de buscar un circuito."""
    if not adjacency:
        return False
    visited, pending = {next(iter(adjacency))}, [next(iter(adjacency))]
    while pending:
        current = pending.pop()
        for neighbor, _ in adjacency[current]:
            if neighbor not in visited:
                visited.add(neighbor)
                pending.append(neighbor)
    return len(visited) == len(adjacency)


def evaluate_criteria(adjacency: dict[str, list[tuple[str, float]]]) -> dict:
    """Evalúa grado mínimo, Dirac y Ore; Dirac/Ore son suficientes, no obligatorios."""
    size = len(adjacency)
    degrees = {node: len(neighbors) for node, neighbors in adjacency.items()}
    minimum = min(degrees.values(), default=0)
    ore = size >= 3
    for node, neighbors in adjacency.items():
        adjacent = {neighbor for neighbor, _ in neighbors}
        for other in adjacency:
            if other != node and other not in adjacent and degrees[node] + degrees[other] < size:
                ore = False
    return {
        "n": size,
        "connected": is_connected(adjacency),
        "degrees": degrees,
        "minimum_degree": minimum >= 2,
        "dirac": size >= 3 and minimum >= size / 2,
        "ore": ore,
    }


def find_hamiltonian_cycle(block: dict, start_id: str) -> dict | None:
    """Busca por retroceso el ciclo válido con menor peso; nunca usa una ruta fija."""
    adjacency = build_adjacency(block)
    if start_id not in adjacency or len(adjacency) < 3:
        return None
    best_weight, best_path = inf, None
    visited, path = {start_id}, [start_id]

    def search(current: str, total: float) -> None:
        nonlocal best_weight, best_path
        if len(path) == len(adjacency):
            closing = next((weight for neighbor, weight in adjacency[current] if neighbor == start_id), None)
            if closing is not None and total + closing < best_weight:
                best_weight, best_path = total + closing, [*path, start_id]
            return
        for neighbor, weight in sorted(adjacency[current], key=lambda item: item[1]):
            if neighbor not in visited and total + weight < best_weight:
                visited.add(neighbor)
                path.append(neighbor)
                search(neighbor, total + weight)
                path.pop()
                visited.remove(neighbor)

    search(start_id, 0)
    return None if best_path is None else {"path": best_path, "total_weight": round(best_weight, 2)}


def analyze_block(block: dict, start_id: str) -> dict:
    adjacency = build_adjacency(block)
    return {"criteria": evaluate_criteria(adjacency), "cycle": find_hamiltonian_cycle(block, start_id)}
